### PR TITLE
color=always in Emacs

### PR DESCRIPTION
**Pull request by fommil**
As of https://github.com/haskell/haskell-mode/pull/1608 (and https://github.com/haskell/haskell-mode/pull/1609) Emacs is able to handle colour output in tasty tests by default.

Before I progress with this PR, I wanted to check if it's something you'd be interested in accepting. It saves me having to add `--color=always` to every invocation.

It is relatively easy to detect that we are within Emacs, we just look for the envvar `INSIDE_EMACS`. The logic to match how this is done in the JVM world is at https://github.com/jline/jline2/pull/288 / https://github.com/jline/jline2/commit/5366158937c49b6e277345e3e543ba61b6fe33b4

TL;DR emacs supports ansi colours but not echo and navigation, when compared to other "smart" terminals.

The situation I'd be concerned about is when cabal runs multiple tasty jobs in a batch and puts their output into a file instead of the console. In such cases I don't think we want colour output... but we would have the same problem in other terminals also. Is there a way to detect that?

A user or project level `.tasty` containing some options would be another way to do this.

### CODE REVIEW

**Comment by feuerbach** [comment]
> TL;DR emacs supports ansi colours but not echo, when compared to other "smart" terminals.

Could you elaborate on what "but not echo" means?

**Comment by feuerbach** [comment]
I'm all for solving the issue that you have with Emacs, but I'm not very comfortable with adding Emacs-specific code to tasty—and this has almost nothing to do with the fact that I'm a vim user ;-)

I feel like the right place to add Emacs-specific checks may be in ansi-terminal, or maybe try to make use of terminfo (but would that work on Windows?). Then we can adjust tasty w.r.t. more abstract capabilities that a terminal provides, instead of checking for Emacs specifically.

**Comment by fommil** [comment]
I don't think echo is relevant for tasty. It's when the user inputs commands, echo would then add them to stdout.

Unfortunately it's an all-or-nothing check in `ansi-terminal` http://hackage.haskell.org/package/ansi-terminal-0.8.2/docs/System-Console-ANSI.html#g:11 whereas Emacs can handle colours but not necessarily the interactive features like cursor navigation. What we need is a `hSupportsANSIColor` and Emacs can return `True`. Would you be open to a PR adding that kind of more refined check?

**Comment by feuerbach** [comment]
> Would you be open to a PR adding that kind of more refined check (to ansi-terminal)?

Yes, I think this would be the right approach.

**Comment by fommil** [comment]
@feuerbach ok, I sent a PR to the upstream lib.

**Comment by fommil** [comment]
To get this to work while the upstream is unreleased add this to `cabal.project`

```
source-repository-package
    type: git
    location: https://github.com/fommil/ansi-terminal.git
    tag: d25db05b900db889dddc91e93cea47c79621225a
```

**Comment by fommil** [comment]
@feuerbach now that `ansi-terminal` 0.9 is released, I updated my PR :smile:

**Comment by feuerbach** [comment]
But isTerm isn't just used to output colors; it's also used to erase stuff. So you may need to distinguish those two capabilities.

**Comment by fommil** [comment]
ok, I'll take a look at that soon.

**Comment by fommil** [comment]
@feuerbach like this?

**Comment by fommil** [comment]
`--hide-successes` works fine in Emacs, and I'm seeing colours.

I pushed an update to the `stack.yaml` so the windows build will hopefully go green.

**Comment by feuerbach** [comment]
Oops, looks like I broke the CI on master. Should be fixed now. Could you please rebase?

**Comment by fommil** [comment]
done

**Comment by feuerbach** [comment]
Thanks, merged and released as tasty-1.2.1.
